### PR TITLE
DM-36167

### PR DESCRIPTION
**Pull request by pkubanek**
Proper time synchronization using NI-9469 modules.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,94 +1,206 @@
...
-As Camera Rotator uses NI-9076 cRIO VxWorks RTOS, C++ based Linux CsC cannot be
-run directly on CPU attached to FPGA. FPGA can be commanded through Ethernet,
-with RIO name specified as URL (rio://<ip-address>/RIO). Please see [TS VMS
-configuration](https://github.com/lsst-ts/ts_vms/master/SettingFiles/CameraRotator/VMSApplicationSettings.yaml)
-for details.
+# Telescope & Site VMS FPGA
+
+This is code for FPGAs loaded into cRIO FPGAs. Handles low-level tasks of
+recording data into FPGA FIFOs. [TS VMS](https://github.com/lsst-ts/ts_vms) is
+a CsC, asking FPGA for data and making those available as SAL/DDS messages.
```
Comment (teweitsai) [inline README.md]: CsC --> CSC

File: README.md [hunk 2]
```diff
@@ -1,94 +1,206 @@
...
+### Response to command 1
+
+:warning: Before reading data, you shall:
+
+1. synchronize FPGA clocks (command 2)
+2. start accelerometers (command 1)
+
+The process requesting data shall regularly call command 2 to synchronize
+clocks as it reads data. Those are internal FPGA clocks, doesn't affect readout
+frequency - it is use only for timestamping samples.
```
Comment (teweitsai) [inline README.md]: it is used only

File: README.md [hunk 3]
```diff
@@ -1,94 +1,206 @@
...
+cRIOs are equipped with NI-9469 synchronization module. One of the VMS cRIOs
+is designated as source of the synchronization signals, generating triggers at
+regular intervals. This source cRIO is refered to as controller in the
+following paragraphs. The other cRIOs are designated as responders. Responders
+receive clock triggers from the controller. NI-9469 provides 4 trigger lines,
+which are used as following:
+
+| Line | Used for                                                                                                                       |
+| ---- | ------------------------------------------------------------------------------------------------------------------------------ |
+| 0    | Triggered for controller-responder synchronization. Data shall be avereged out, base clock increased when this trigger occurs. |
```
Comment (teweitsai) [inline README.md]: averaged

File: README.md [hunk 4]
```diff
@@ -1,94 +1,206 @@
...
+  controller-->Fail[Synchronization Failed]
+  responder-->Fail[Synchronization Failed]
+  subgraph init[" "]
+  Fail-->Initialising
+  end
+```
+
+### Initialising state
+
+When CSC is started, it signals in its state it is initialising. In this state,
```
Comment (teweitsai) [inline README.md]: "it signals in its state it is initialising" --> I do not understand this sentence.

File: README.md [hunk 5]
```diff
@@ -1,94 +1,206 @@
...
+Synced state, or when any responder transition into Synchronization Failed
+state. 
+
+Responder should switch to this state if triggers on line 0 aren't received for
+given multiply of expected intervals. Of course responder switch to this state
+when controller switch to this state.
+
+Once received, controller CSC shall stop sending triggers, and both controller
+and responder CSC shall stop measuring loop. After given time delay, the CSC
+transitions into initialising state.
```
Comment (teweitsai) [inline README.md:189]: initialising --> initializing

File: README.md [hunk 6]
```diff
@@ -1,94 +1,206 @@
...
+
+This state can be used by a CSC to force base offset resets and time
+synchronization.
+
+## Data processing
+
+Running statistics (mean, average,  minimal, maximal values) is calculated with
+every new sample, acquired from AD converter. Individual readout values are
+also stored in a memory, allowing calculation of mean, if needed.
```
Comment (teweitsai) [inline README.md]: in the memory

**Review by teweitsai — APPROVED** [review]
Looks good. I confess I did not check the details of vi, but I suppose they should work correctly.

File: README.md (hunk 4 above)
Comment (pkubanek) [reply README.md]: CSC is started in initialising state.

**Pull request merged**